### PR TITLE
`Programming exercises`: Improve error handling when using the online code editor

### PR DESCRIPTION
**Pull request by dfuchss**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

I've encountered multiple internal server errors on our production system regarding git.

### Description

I've added error handlers for those errors I've encountered.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

Test that online editor still works (starting, pushing)

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [ ] Test

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 1]
```diff
@@ -518,6 +530,8 @@ public Path getDefaultLocalPathOfRepo(VcsRepositoryUrl targetUrl) {
      * @return path of the local file system
      */
     public Path getLocalPathOfRepo(String targetPath, VcsRepositoryUrl targetUrl) {
+        if (targetUrl == null)
+            return null;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:534]:
```suggestion
        if (targetUrl == null) {
            return null;
        }
```
The code style checker probably wants braces here.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 2]
```diff
@@ -491,6 +501,8 @@ private void waitUntilPathNotBusy(final Path localPath) throws CanceledException
      */
     public boolean isRepositoryCached(VcsRepositoryUrl repositoryUrl) {
         Path localPath = getLocalPathOfRepo(repoClonePath, repositoryUrl);
+        if (localPath == null)
+            return false;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:505]:
```suggestion
        if (localPath == null) {
            return false;
        }
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 3]
```diff
@@ -420,7 +417,18 @@ public Repository getOrCheckoutRepository(VcsRepositoryUrl sourceRepoUrl, VcsRep
...
-                pull(repository);
+                try {
+                    pull(repository);
+                }
+                catch (JGitInternalException | NoHeadException | TransportException e) {
+                    // E.g., LockFailedException
+                    // cleanup the folder to avoid problems in the future.
+                    // 'deleteQuietly' is the same as 'deleteDirectory' but is not throwing an exception, thus we avoid another try-catch block.
+                    if (!FileUtils.deleteQuietly(localPath.toFile())) {
+                        log.error("Could not delete directory during pull: {}", localPath.toAbsolutePath());
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:428]:
```suggestion
                        log.error("Could not delete directory after failed pull: {}", localPath.toAbsolutePath());
```

File: src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java [hunk 4]
```diff
@@ -441,7 +449,9 @@ public Repository getOrCheckoutRepository(VcsRepositoryUrl sourceRepoUrl, VcsRep
             catch (IOException | URISyntaxException | GitAPIException | InvalidPathException e) {
                 // cleanup the folder to avoid problems in the future.
                 // 'deleteQuietly' is the same as 'deleteDirectory' but is not throwing an exception, thus we avoid another try-catch block.
-                FileUtils.deleteQuietly(localPath.toFile());
+                if (!FileUtils.deleteQuietly(localPath.toFile())) {
+                    log.error("Could not delete directory during clone: {}", localPath.toAbsolutePath());
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/GitService.java:453]:
```suggestion
                    log.error("Could not delete directory after failed clone: {}", localPath.toAbsolutePath());
```

**Review by b-fein — APPROVED** [review]
Small suggestions, otherwise nice fixes for some errors I’ve also seen before.

Manual testing is not really possible as all those errors only happen sporadically when something in the communication with the VCS and/or the filesystem fails.

**Comment by dfuchss** [comment]
> Small suggestions, otherwise nice fixes for some errors I’ve also seen before.
> 
> Manual testing is not really possible as all those errors only happen sporadically when something in the communication with the VCS and/or the filesystem fails. As this only improves the clean-up of faulty files for Artemis, it should not cause any problems with the online editor and/or programming exercise start.

Thanks for the suggestions. I've applied all of them.

**Review by b-fein — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code changes look good to me

**Review by timo07k — APPROVED** [review]
Works as expected, online editor still works. Tested on TS1.

**Review by Strohgelaender — APPROVED** [review]

**Pull request merged**